### PR TITLE
fix(nodes, test): scope core-module stubs so they don't leak across tests

### LINKED ISSUE #1640: Test pollution: node tests leak core-module stubs (rocketlib/pydantic) into sys.modules
## Summary

Several node tests install stubs for **core modules** (`rocketlib` and its submodules, `pydantic`) into `sys.modules` at **module import time** and never restore them. The stub then stays in `sys.modules` for the rest of the pytest session and **leaks into other tests** that import the real module. The suite currently passes only because the collection/execution order happens to keep each polluting test away from the tests that need the real module. Any change that reorders collection (e.g. renaming node folders, adding a test, a different xdist split) can surface it.

## Evidence

Confirmed while renaming node folders (which changes pytest collection order):

- `nodes/test/atlas/test_isdeleted_filter.py` installed a partial `pydantic` stub (`ValidationError` only, no `BaseModel`) with `sys.modules.setdefault('pydantic', ...)` at module level, never restored. When it ran before the chroma/weaviate tests, those failed with:
  ```
  ImportError: cannot import name 'BaseModel' from 'pydantic' (unknown location)
  ```
- `nodes/test/cloud_tts/test_cloud_tts.py` installs `rocketlib.OPEN_MODE = type('OPEN_MODE', (), {'CONFIG': 'config'})` — an `OPEN_MODE` with only `CONFIG`, no `SOURCE` — into a `rocketlib` stub that leaks. Tests that import the real `rocketlib` then fail with:
  ```
  AttributeError: type object 'OPEN_MODE' has no attribute 'SOURCE'
  ImportError: cannot import name 'AVI_ACTION' from 'rocketlib' (unknown location)
  ```
  Observed failures in `tool_slack`, `rerank_cohere`, `face_detection` — none of which install the stub.

The `(unknown location)` in these errors is the signature of a hand-built `ModuleType` sitting in `sys.modules` in place of the real package.

## Root cause

Module-level `sys.modules['<core>'] = stub` (or `setdefault`) with no teardown. Because it runs during collection (import of the test file), it is global and persists across the session, and under `pytest-xdist` it contaminates whatever else lands in the same worker.

## Affected files (to audit)

Tests that assign a core module into `sys.modules` at module level — each needs checking for whether it restores:

- `nodes/test/cloud_tts/test_cloud_tts.py`
- `nodes/test/guardrails/test_all.py`
- `nodes/test/ocr/test_model_server_ocr.py`
- `nodes/test/test_db_hydradb.py`
- `nodes/test/test_vectordb_tool_mixin.py`
- `nodes/test/tool_filesystem/test_read_size_cap.py`
- `nodes/test/tool_mem0/test_tools.py`
- `nodes/test/tool_xtrace_memory/test_tools.py`
- `nodes/test/test_anomaly_detector.py`
- `nodes/test/test_currency_convert_explicit.py`
- `nodes/test/test_baidu_qianfan_global_validation.py`
- `nodes/test/test_graph_falkordb.py`, `nodes/test/test_graph_neo4j.py`
- `nodes/test/test_contracts.py`, `nodes/test/test_tool_daytona.py`
- `nodes/test/agent_crewai/test_stop_words.py` (appears to already restore — verify)

(`nodes/test/atlas/test_isdeleted_filter.py` is already fixed as the reference.)

## Proposed fix

Scope every core-module stub to the node import that needs it and restore `sys.modules` afterwards, so it never leaks:

```python
_saved = sys.modules.get('rocketlib')
stub = types.ModuleType('rocketlib')
stub.OPEN_MODE = ...  # complete enough for the import
sys.modules['rocketlib'] = stub
try:
    mod = importlib.import_module('nodes.<node>.IGlobal')
finally:
    if _saved is not None:
        sys.modules['rocketlib'] = _saved
    else:
        sys.modules.pop('rocketlib', None)
```

Restore each stubbed module (including submodules like `rocketlib.types`). This makes the suite order-independent and unblocks work that legitimately reorders collection (e.g. node renames).

## Why now

The node-family rename PR (#1636) is blocked by this: it does nothing to these tests, but its folder renames reorder collection and expose the leak. Fixing isolation here, on its own PR, lets that PR (and any future reorder) go green.

### PR DESCRIPTION
**Pull request by dsapandora**
## Summary

- Several node tests installed stubs for core modules (`rocketlib` + submodules, `pydantic`, and their `ai.common.*`/`requests`/`tenacity` companions) into `sys.modules` at import time and never restored them. The stub stayed global for the rest of the pytest session and leaked into unrelated tests that import the real module.
- Wrap each stub so it lives only while the node code is imported, and restore `sys.modules` in a `finally`. Switch `setdefault` to direct assignment so a stub left contaminated by another test can't be reused.
- No test logic or assertions changed — only the stub lifecycle.

## Type

test / bug fix (test isolation)

## Testing

- [x] Tests updated (stub save/restore added; behaviour unchanged)
- [ ] Tested locally (lint green: ruff; full suite runs in CI — no engine locally)
- [ ] `./builder test` passes (runs in CI on this PR)

## Checklist

- [x] Commit messages follow [conventional commits](https://www.conventionalcommits.org/)
- [x] No secrets or credentials included
- [ ] Wiki updated (if applicable) — N/A
- [x] Breaking changes documented — none; test-only, no runtime behaviour changes

## Linked Issue

Fixes #1640

<sub>Context: this unblocks #1636 (node-family renames), which does nothing to these tests but reorders pytest collection and surfaces the leak. With isolation fixed here, any reorder stays green.</sub>

## Summary by CodeRabbit

* **Bug Fixes / Test Reliability**
  * Improved test isolation by scoping dependency stubs to each test’s import window and restoring prior `sys.modules` state afterward.
  * Applied a consistent save-install-restore pattern across node and tool tests to prevent mock/module leakage between tests.
  * Updated dynamic imports to avoid persistent stub injection while keeping existing assertions unchanged.
  * Added an automatic pytest regression guard that detects leaked stubbed core modules and fails the test to enforce correct isolation.

### CODE REVIEW

File: nodes/test/tool_mem0/test_tools.py [hunk 1]
```diff
@@ -126,21 +126,42 @@ def _ensure_pkg() -> None:
...
+    'requests.exceptions',
+    'tenacity',
+)
+_saved_core = {_name: sys.modules.get(_name) for _name in _CORE_STUBS}
+
 _ensure_rocketlib()
 _ensure_ai_common()
 _ensure_requests()
 _ensure_tenacity()
 _ensure_pkg()
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/tool_mem0/test_tools.py:140]: Thanks, but this doesn't apply here. This PR's whole point is that every affected test **snapshots** the core modules it stubs (`_saved_core`) and **restores** `sys.modules` in a `finally` after the node import — so no test leaves a contaminated stub for a later one to reuse. There's nothing left to reuse.

The `_ensure_*` helpers are intentionally *augment, never clobber*. When the real `rocketlib` is loaded it already exports everything they'd add (`IInstanceBase`, `IGlobalBase`, `tool_function`, `OPEN_MODE`, `debug/error/warning`), so they add and mutate nothing. Assigning a fresh `ModuleType` unconditionally would instead replace the real module with an incomplete stub (missing `AVI_ACTION`, `BaseModel` via `rocketlib.types`, …) and break transitive imports — the exact failure this PR fixes in the atlas test. `cloud_tts` uses direct assignment because it never loads the real rocketlib; `tool_mem0`/`tool_xtrace` coexist with it. All three OS builds are green.

[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/tool_mem0/test_tools.py:140]: Good eye on the mutation angle — but it doesn't bite here. With the real `rocketlib` loaded, it already exports every symbol `_ensure_rocketlib` would add (`OPEN_MODE`, `IInstanceBase`, `tool_function`, `debug/error/warning`), so every `hasattr` is true and the helper mutates **nothing**. And since this PR makes every rocketlib-stubbing test restore `sys.modules` in a `finally`, no sibling test leaves a stub object behind for this file to mutate — the pre-existing-stub case you describe no longer occurs.

Building a fresh `ModuleType` unconditionally would be the riskier move: it replaces the real module with an incomplete stub (no `AVI_ACTION`, no `BaseModel` via `rocketlib.types`) and breaks transitive imports — precisely the failure this PR fixes in atlas. Verified green on Ubuntu/macOS/Windows.

File: nodes/test/tool_xtrace_memory/test_tools.py [hunk 2]
```diff
@@ -117,15 +117,36 @@ def _ensure_pkg() -> None:
...
+    'requests',
+    'requests.exceptions',
+    'tenacity',
+)
+_saved_core = {_name: sys.modules.get(_name) for _name in _CORE_STUBS}
+
 _ensure_rocketlib()
 _ensure_ai_common()
 _ensure_requests()
 _ensure_tenacity()
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/tool_xtrace_memory/test_tools.py:130]: Same as the tool_mem0 thread: with the real `rocketlib` present it already exports everything `_ensure_rocketlib` would add, so nothing is mutated; and every rocketlib-stubbing test in this PR restores `sys.modules` in a `finally`, so no sibling stub is left to mutate. A fresh unconditional `ModuleType` would replace the real module with an incomplete stub and break `rocketlib.types` imports (`AVI_ACTION`/`BaseModel`) — the atlas failure this PR fixes. Green on all three OS.

**Review by asclearuc — DISMISSED** [review]

File: nodes/test/tool_mem0/test_tools.py [hunk 3]
```diff
@@ -71,29 +68,24 @@ def _passthrough(args, tool_name=None):
...
-    """Create minimal ``ai.common.*`` stubs only when absent (never overwrite)."""
+    """Install fresh ``ai.common.*`` stubs so the node imports without the engine."""
     for name in ('ai', 'ai.common', 'ai.common.utils', 'ai.common.config'):
-        if name not in sys.modules:
-            sys.modules[name] = types.ModuleType(name)
-    if not hasattr(sys.modules['ai.common.utils'], 'normalize_tool_input'):
-        sys.modules['ai.common.utils'].normalize_tool_input = _passthrough
-    if not hasattr(sys.modules['ai.common.config'], 'Config'):
+        sys.modules[name] = types.ModuleType(name)
+    sys.modules['ai.common.utils'].normalize_tool_input = _passthrough
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/tool_mem0/test_tools.py:77]: Done in `f2e7560d` — the `ai`/`ai.common` stubs are now marked as packages (`__path__ = []`), matching the ocr/face_detection pattern so `ai.common.*` sub-imports resolve.

File: nodes/test/tool_xtrace_memory/test_tools.py [hunk 4]
```diff
@@ -62,45 +59,38 @@ def _passthrough(args, tool_name=None):
...
+    below); the fresh stubs never leak thanks to the _saved_core restore.
     """
     for name in ('ai', 'ai.common', 'ai.common.utils', 'ai.common.config'):
-        if name not in sys.modules:
-            sys.modules[name] = types.ModuleType(name)
-    if not hasattr(sys.modules['ai.common.utils'], 'normalize_tool_input'):
-        sys.modules['ai.common.utils'].normalize_tool_input = _passthrough
-    if not hasattr(sys.modules['ai.common.config'], 'Config'):
+        sys.modules[name] = types.ModuleType(name)
+    sys.modules['ai.common.utils'].normalize_tool_input = _passthrough
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/tool_xtrace_memory/test_tools.py:72]: Done in `f2e7560d` — the `ai`/`ai.common` stubs are now marked as packages (`__path__ = []`), matching the ocr/face_detection pattern so `ai.common.*` sub-imports resolve.

File: nodes/test/cloud_tts/test_cloud_tts.py [hunk 5]
```diff
@@ -31,30 +31,41 @@
...
+    sys.modules['rocketlib'] = rocketlib
 
-    sys.modules.setdefault('ai', types.ModuleType('ai'))
-    sys.modules.setdefault('ai.common', types.ModuleType('ai.common'))
+    sys.modules['ai'] = types.ModuleType('ai')
+    sys.modules['ai.common'] = types.ModuleType('ai.common')
     ai_cfg = types.ModuleType('ai.common.config')
     ai_cfg.Config = type('Config', (), {})
-    sys.modules.setdefault('ai.common.config', ai_cfg)
+    sys.modules['ai.common.config'] = ai_cfg
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/cloud_tts/test_cloud_tts.py:52]: Done in `f2e7560d` — the `ai`/`ai.common` stubs are now marked as packages (`__path__ = []`), matching the ocr/face_detection pattern so `ai.common.*` sub-imports resolve.

File: nodes/test/test_db_hydradb.py [hunk 6]
```diff
@@ -164,8 +163,16 @@ def node():
...
+    ]
+    _saved = {name: sys.modules.get(name) for name in _managed}
+    Config = _install_stubs()
+    client_mod, iglobal_mod, iinstance_mod = _load_package()
+    yield SimpleNamespace(Config=Config, client_mod=client_mod, iglobal_mod=iglobal_mod, iinstance_mod=iinstance_mod)
+    for name in _managed:
+        if _saved[name] is not None:
+            sys.modules[name] = _saved[name]
+        else:
+            sys.modules.pop(name, None)
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/test_db_hydradb.py]: Done in `f2e7560d` — the fixture now wraps setup + yield + restore in `try/finally`, so `sys.modules` is restored even if setup raises.

File: nodes/test/qdrant/test_search_threshold.py [hunk 7]
```diff
@@ -38,17 +38,28 @@ def doesCollectionExist(self, *a, **kw):
...
-_spec.loader.exec_module(_qdrant_mod)
-Store = _qdrant_mod.Store
+)
+
+# Snapshot every name we touch (setdefault set + ai.common.store), then restore
+# in finally so these stubs never leak to sibling tests run in the same process.
+_saved = {_name: sys.modules.get(_name) for _name in (*_STUB_NAMES, 'ai.common.store')}
+try:
+    for _name in _STUB_NAMES:
+        sys.modules.setdefault(_name, MagicMock())
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/qdrant/test_search_threshold.py]: Done in `f2e7560d` — switched to direct assignment (`sys.modules[_name] = MagicMock()`); the surrounding save/restore already prevents leaks.

File: nodes/test/test_anomaly_detector.py [hunk 8]
```diff
@@ -10,12 +10,20 @@
...
 
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src', 'nodes', 'anomaly_detector'))
-from detector import AnomalyDetector
+try:
+    from detector import AnomalyDetector
+finally:
+    if _saved_rl is not None:
+        sys.modules['rocketlib'] = _saved_rl
+    else:
+        sys.modules.pop('rocketlib', None)
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/test_anomaly_detector.py:26]: Done in `f2e7560d` — added `# noqa: E402`. Note E402 is already in ruff's ignore list, so this is for consistency with sibling tests rather than a lint fix.

File: nodes/test/conftest.py [hunk 9]
```diff
@@ -50,6 +50,55 @@
...
+        return True
+    # Real modules and namespace packages carry a __spec__ or __file__; bare
+    # types.ModuleType stubs built in tests carry neither.
+    return getattr(mod, '__spec__', None) is None and getattr(mod, '__file__', None) is None
+
+
+# Module-scoped: checked once per test module, after its module-scoped fixtures
+# have torn down. A per-test check would false-positive on tests that legitimately
+# stub a core module in a scope='module' fixture and restore it at module teardown.
+@pytest.fixture(scope='module', autouse=True)
```
Comment (asclearuc) [inline nodes/test/conftest.py]: **The guard blames the wrong file, and one leak cascades into N failures.**

This is a module-scoped fixture, so it checks at **teardown**. But the leaks it targets happen at **import time**, and pytest imports every test module during collection — before any test runs, long before any teardown. By the time the first module tears down, a leak from a module collected much later is already in `sys.modules`. The guard sees it and blames whichever module happens to be finishing. It also never cleans up, so the same leak re-fires at every later module teardown.

I verified this. One deliberate leaker at `nodes/test/zz_leaker/test_leak.py` (3 lines: stub `rocketlib` at import, never restore), run with five modules this PR already fixed:

```
129 passed, 6 errors

ERROR nodes/test/cloud_tts/test_cloud_tts.py::...
ERROR nodes/test/guardrails/test_all.py::...
ERROR nodes/test/qdrant/test_search_threshold.py::...
ERROR nodes/test/tool_mem0/test_tools.py::...
ERROR nodes/test/test_anomaly_detector.py::...
ERROR nodes/test/zz_leaker/test_leak.py::...
```

One leaker, six errors. Five name modules that restore correctly — four of them files you just fixed in this PR — each saying:

> `nodes/test/cloud_tts/test_cloud_tts.py left stub core modules in sys.modules: ['rocketlib']. Snapshot and restore sys.modules around the node import (save/restore pattern).`

`cloud_tts` already does exactly that. The message tells the reader to apply a fix that is already applied, in a file that did nothing wrong.

**Heads-up: snapshotting in the fixture's setup will not fix it.** Module-scoped setup also runs after collection, so the offender's own leak is already in its baseline and it would never be caught. The check has to run at collection time, right after each module is imported.

Here is a version I prototyped and ran against the same set:

```python
_core_baseline = {}
_core_leaks = []


def pytest_collectstart(collector):
    """Snapshot the guarded modules just before a test module is imported."""
    if type(collector).__name__ == 'Module':
        _core_baseline[collector.nodeid] = {n: sys.modules.get(n) for n in _GUARDED_CORE_MODULES}


def pytest_collectreport(report):
    """Right after a test module is imported, record + undo any stub it left behind."""
    base = _core_baseline.pop(report.nodeid, None)
    if base is None:
        return
    leaked = [
        n
        for n in _GUARDED_CORE_MODULES
        if sys.modules.get(n) is not base[n] and _is_stub_module(sys.modules.get(n))
    ]
    if not leaked:
        return
    _core_leaks.append((report.nodeid, leaked))
    # Undo the leak so one offender does not cascade into every later module.
    for n in _GUARDED_CORE_MODULES:
        if base[n] is None:
            sys.modules.pop(n, None)
        else:
            sys.modules[n] = base[n]


def pytest_terminal_summary(terminalreporter):
    """Report leaked core-module stubs (see #1640)."""
    for nodeid, leaked in _core_leaks:
        terminalreporter.write_line(
            f'sys.modules leak: {nodeid} left stub core modules {leaked}. '
            'Snapshot and restore sys.modules around the node import.',
            red=True,
        )


def pytest_sessionfinish(session, exitstatus):  # noqa: ARG001
    """Fail the run when any test module leaked a core-module stub."""
    if _core_leaks:
        session.exitstatus = 1
```

Same experiment against this version:

```
129 passed

sys.modules leak: nodes/test/zz_leaker/test_leak.py left stub core modules ['rocketlib'].
                  Snapshot and restore sys.modules around the node import.
```

One message, correct file, no false positives. Exit code `1` with the leaker present, `0` with it removed.

One gotcha: do not `raise` from inside the hook — that gives pytest `INTERNALERROR` and kills the run before any test executes. Record and fail at session finish.

Also please add the negative test: a fixture file that deliberately leaks, plus an assertion that the guard names *it*. The guard has no test of its own right now, and that is exactly what catches this.

File: nodes/test/conftest.py [hunk 10]
```diff
@@ -50,6 +50,55 @@
...
+    'ai.common.schema',
+    'ai.common.store',
+    'pydantic',
+    'requests',
+    'requests.exceptions',
+    'tenacity',
+)
+
+
+def _is_stub_module(mod) -> bool:
```
Comment (asclearuc) [inline nodes/test/conftest.py]: Minor, on the detection heuristic: "no `__spec__` and no `__file__`" misses a stub built with `importlib.util.module_from_spec`, which carries a `__spec__`. Several tests in this repo build stubs that way.

If you move to the per-module baseline suggested in the other comment, this stops being the only signal — you compare identity against the snapshot first, and this heuristic just filters out legitimate real-module swaps. That closes the gap without needing a smarter heuristic.

File: nodes/test/test_anomaly_detector.py [hunk 11]
```diff
@@ -10,12 +10,20 @@
...
 import types
 
+# Stub rocketlib only while importing detector, then restore so it never leaks.
+_saved_rl = sys.modules.get('rocketlib')
 rocketlib = types.ModuleType('rocketlib')
 rocketlib.debug = lambda *a, **kw: None
-sys.modules.setdefault('rocketlib', rocketlib)
+sys.modules['rocketlib'] = rocketlib
 
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src', 'nodes', 'anomaly_detector'))
```
Comment (asclearuc) [inline nodes/test/test_anomaly_detector.py:19]: Non-blocking, follow-up: `sys.modules['rocketlib']` is restored now, but this `sys.path.insert` stays for the whole session, and so does `sys.modules['detector']` after the import below.

`detector` is a very generic name in a shared namespace. If a second node ever ships a `detector.py`, a later test doing its own `sys.path.insert` + `from detector import ...` gets this one back out of `sys.modules` and silently tests the wrong module. Same shape of bug as the one this PR fixes, just latent — today `anomaly_detector/detector.py` is the only file with that name.

Loading by explicit file path (`spec_from_file_location`) the way `guardrails/test_all.py` does avoids both the `sys.path` entry and the generic name.

File: nodes/test/test_currency_convert_explicit.py [hunk 12]
```diff
@@ -10,13 +10,20 @@
...
 # convert.py has no rocketlib import, but stub it defensively so importing the
-# node package stays isolated from the native engine.
+# node package stays isolated; restore after so the stub never leaks.
+_saved_rl = sys.modules.get('rocketlib')
 rocketlib = types.ModuleType('rocketlib')
 rocketlib.debug = lambda *a, **kw: None
-sys.modules.setdefault('rocketlib', rocketlib)
+sys.modules['rocketlib'] = rocketlib
 
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src', 'nodes', 'currency_convert_explicit'))
```
Comment (asclearuc) [inline nodes/test/test_currency_convert_explicit.py:19]: Same as `test_anomaly_detector.py`: the `sys.path.insert` and `sys.modules['convert']` outlive the import. `convert` is generic enough to collide later. Non-blocking follow-up.

**Review by asclearuc — CHANGES_REQUESTED** [review]

File: nodes/test/conftest.py (hunk 9 above)
Comment (dsapandora) [reply nodes/test/conftest.py]: Fixed in `51703eb3` — thanks for the reproduction. Rewrote it as collection-time hooks: `pytest_collectstart` snapshots the guarded modules before each module import, `pytest_collectreport` compares right after and undoes the leak so it cannot cascade, failing at `pytest_sessionfinish` (no `raise` in the hook). Moved it to `_sys_modules_guard.py` and added `test_sys_modules_guard.py` — the negative test asserts the guard names the real leaker, undoes it, and does not blame an earlier clean module.

File: nodes/test/conftest.py (hunk 10 above)
Comment (dsapandora) [reply nodes/test/conftest.py]: Done — identity vs the per-module snapshot is now the primary signal; `_is_stub_module` just filters legitimate real-module swaps, keyed off `__file__`/`__path__` so it also catches a `module_from_spec` stub. Unit-tested.

File: nodes/test/test_anomaly_detector.py (hunk 11 above)
Comment (dsapandora) [reply nodes/test/test_anomaly_detector.py:19]: Agreed — the generic `detector` name in a shared namespace is a latent trap. Tracking it as the non-blocking follow-up you flagged: load via `spec_from_file_location` (like guardrails) so neither the `sys.path` entry nor `sys.modules[\"detector\"]` outlives the import. Not in this PR.

File: nodes/test/test_currency_convert_explicit.py (hunk 12 above)
Comment (dsapandora) [reply nodes/test/test_currency_convert_explicit.py:19]: Same as the anomaly thread — `convert` is generic enough to collide later. Tracked as a non-blocking follow-up to load via `spec_from_file_location`.

File: nodes/test/tool_mem0/test_tools.py [hunk 13]
```diff
@@ -33,7 +33,9 @@
 
 
 # ---------------------------------------------------------------------------
-# Composable import scaffolding (augments existing stubs, never clobbers)
+# Import scaffolding: build FRESH stubs unconditionally. Never read the object
+# already in sys.modules — the _saved_core save/restore below repones the real
+# one, so a fresh stub can't leak even without the engine loaded.
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/tool_mem0/test_tools.py:38]: Applied — thanks; the comment now reads "restores".

File: nodes/test/tool_xtrace_memory/test_tools.py [hunk 14]
```diff
@@ -29,7 +29,9 @@
 
 
 # ---------------------------------------------------------------------------
-# Composable import scaffolding (augments existing stubs, never clobbers)
+# Import scaffolding: build FRESH stubs unconditionally. Never read the object
+# already in sys.modules — the _saved_core save/restore below repones the real
+# one, so a fresh stub can't leak even without the engine loaded.
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/tool_xtrace_memory/test_tools.py:34]: Applied — thanks; the comment now reads "restores".

File: nodes/test/_sys_modules_guard.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+            f'sys.modules leak: {nodeid} left stub core modules {leaked}. '
+            'Snapshot and restore sys.modules around the node import (save/restore pattern).',
+            red=True,
+        )
+
+
+def pytest_sessionfinish(session, exitstatus):  # noqa: ARG001
+    """Fail the run when any test module leaked a core-module stub."""
+    if _core_leaks:
+        session.exitstatus = 1
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/_sys_modules_guard.py]: Good point — fixed in `049952dd`: `pytest_sessionfinish` now only forces exit 1 when the run was otherwise passing (`exitstatus == 0`), so it will not mask a more severe INTERNALERROR/usage code.

File: nodes/test/test_sys_modules_guard.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+import _sys_modules_guard as guard
+
+
+class _Module:
+    """Stand-in for pytest's Module collector; only its class name matters to the guard."""
+
+    def __init__(self, nodeid):
+        """Record the collector nodeid."""
+        self.nodeid = nodeid
+
```
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/test_sys_modules_guard.py]: Already fixed in `51703eb3` — the stand-in class is now named `Module`, so `type(collector).__name__ == "Module"` matches and the hook path is actually exercised. (CI caught it before this landed.)

File: nodes/test/_sys_modules_guard.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (dsapandora) [reply nodes/test/_sys_modules_guard.py]: Update: reverted, keeping the guard as-is. This is a trivial nit that does not affect correctness — the guard already fails the run correctly whenever a leak is recorded. The masking scenario (a session already exiting with a more severe code AND a leaked stub in the same run) is a corner case, and in it failing is still the right outcome, so the extra `exitstatus == 0` refinement is not worth its own commit/CI cycle. No behavior change intended.

File: nodes/test/test_anomaly_detector.py (hunk 11 above)
Comment (asclearuc) [inline nodes/test/test_anomaly_detector.py:19]: (Repeat of a non-blocking note from the last round, re-anchored — safe to defer.)

`sys.modules['rocketlib']` is restored now, but this `sys.path.insert` stays for the whole session, and so does `sys.modules['detector']` after the import below. `detector` is a very generic name; if a second node ever ships a `detector.py`, a later test doing its own `sys.path.insert` + `from detector import ...` gets this one back out of `sys.modules` and silently tests the wrong module. Loading by explicit file path (`spec_from_file_location`), the way `guardrails/test_all.py` does, avoids both.

File: nodes/test/test_currency_convert_explicit.py (hunk 12 above)
Comment (asclearuc) [inline nodes/test/test_currency_convert_explicit.py:19]: (Non-blocking, same as `test_anomaly_detector.py`.) The `sys.path.insert` and `sys.modules['convert']` outlive the import; `convert` is generic enough to collide later. Safe to defer.

**Review by asclearuc — APPROVED** [review]
Rechecked at `51703eb3`. The blocking issue from the last round is fully resolved — this is an **Approve** from me.

The guard now runs at collection time (`pytest_collectstart` snapshot + `pytest_collectreport` compare) in a dedicated `_sys_modules_guard.py`, records leaks and fails the run in `pytest_sessionfinish`, and undoes each leak so one offender can't cascade. The detection helper is extracted as `check_after_import` and unit-tested. `_is_stub_module` now also handles the `module_from_spec` case (via `__path__`), and `test_sys_modules_guard.py` covers it explicitly — both things I'd flagged.

I re-ran the same experiment that failed last round: one deliberate import-time leaker among five modules that restore correctly.

- **Before:** 6 errors, five of them blaming innocent files.
- **Now:** the leak is attributed to the leaker alone, no false positives, and the run exits `1`. Removing the leaker → exit `0`. The touched set is 169 passed; the guard's own 4 tests pass.

Two small follow-ups remain (both non-blocking, both flagged inline — fine to defer):

- `test_anomaly_detector.py` / `test_currency_convert_explicit.py` still leave `sys.path.insert(...)` and the generic `sys.modules['detector']` / `['convert']` entries for the session.
- The hand-rolled save/restore is now in 8 files plus the guard; `patch.dict(sys.modules, {...})` would collapse each to one line, and it's already the idiom here (`tool_git`, `milvus`, `text_output`).

Nice work turning the guard around, @dsapandora — the collection-time version plus its own test is exactly right. Approving.

File: nodes/test/test_currency_convert_explicit.py (hunk 12 above)
Comment (dsapandora) [reply nodes/test/test_currency_convert_explicit.py:19]: I will defer this one

File: nodes/test/test_anomaly_detector.py (hunk 11 above)
Comment (dsapandora) [reply nodes/test/test_anomaly_detector.py:19]: I will also defer this one
